<!doctype html>
<html class="no-js" lang="en" data-content_root="./">
  <head><meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="genindex.html"><link rel="search" title="Search" href="search.html"><link rel="prev" title="Installation Guide" href="installation_guide.html">

    <!-- Generated with Sphinx 9.1.0 and Furo 2025.12.19 -->
        <title>RFDpoly usage examples - RFDpoly 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=acfd86a5" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?v=7bdb33bb" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?v=8dab3a3b" />
    
    


<style>
  body {
    --color-code-background: #f2f2f2;
  --color-code-foreground: #1e1e1e;
  --color-brand-primary: #F68A33;
  --color-brand-content: #37939B;
  --font-stack: Open Sans, sans-serif;
  --font-stack--headings: Open Sans, sans-serif;
  --color-background-hover: #DCE8E8ff;
  --color-announcement-background: #F68A33dd;
  --color-announcement-text: #070707;
  --color-brand-visited: #37939B;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  --color-brand-primary: #37939B;
  --color-brand-content: #F68A33;
  --font-stack: Open Sans, sans-serif;
  --font-stack--headings: Open Sans, sans-serif;
  --color-brand-visited: #37939B;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  --color-brand-primary: #37939B;
  --color-brand-content: #F68A33;
  --font-stack: Open Sans, sans-serif;
  --font-stack--headings: Open Sans, sans-serif;
  --color-brand-visited: #37939B;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle site navigation sidebar">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc" aria-label="Toggle table of contents sidebar">
<label class="overlay sidebar-overlay" for="__navigation"></label>
<label class="overlay toc-overlay" for="__toc"></label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <span class="icon"><svg><use href="#svg-menu"></use></svg></span>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">RFDpoly 1.0.0 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  <span class="sidebar-brand-text">RFDpoly 1.0.0 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">readme_link.rst</a></li>
<li class="toctree-l1"><a class="reference internal" href="license_link.html">LICENSE</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing_link.html">CONTRIBUTING to RFDpoly</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation_guide.html">Installation Guide</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">RFDpoly usage examples</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="_sources/rfdpoly_usage_examples.md.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="rfdpoly-usage-examples">
<h1>RFDpoly usage examples<a class="headerlink" href="#rfdpoly-usage-examples" title="Link to this heading">¶</a></h1>
<p>This document has several examples for how to use RFDpoly for RNA and DNA generation including:</p>
<ul class="simple">
<li><p>Unconditional generation of multiple biopolymers together</p></li>
<li><p>Motif scaffolding tasks with multiple polymers</p></li>
<li><p>Control of base-pairing secondary structure to design RNA pseudoknots or (mini) DNA origami</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you come across this error:
<code class="docutils literal notranslate"><span class="pre">bash</span>&#160;&#160;&#160;&#160; <span class="pre">Could</span> <span class="pre">note</span> <span class="pre">locate</span> <span class="pre">file</span> <span class="pre">'/current/working/directory/test_data/DBP035.pdb'.</span> <span class="pre">Tried...</span>&#160;&#160;&#160;&#160; </code>
You can fix it via the addition of
<code class="docutils literal notranslate"><span class="pre">bash</span>&#160;&#160;&#160;&#160; <span class="pre">inference.input_pdb=/path/to/RFDpoly/rf_diffusion/test_data/DBP035.pdb</span>&#160;&#160;&#160;&#160; </code>
to the end of your list of configuration options.</p>
</div>
<section id="table-of-contents">
<h2>Table of Contents<a class="headerlink" href="#table-of-contents" title="Link to this heading">¶</a></h2>
<ul>
<li><p><a class="reference internal" href="#unconditional-rna-generation">Example 1: Unconditional RNA generation</a></p></li>
<li><p><a class="reference internal" href="#unconditional-protein-2-dna">Example 2: Example 2: Unconditional design of one protein chain and two DNA chains</a></p></li>
<li><p><a class="reference internal" href="#rna-riboswitch-conditioning-eterna">Example 3: RNA riboswitch design with conditioning from Eterna puzzle structure string</a></p></li>
<li><p><a class="reference internal" href="#dna-protein-scaffolding-inpaint-2-dna">Example 4: DNA–protein scaffolding, inpaint two DNA chains</a></p></li>
<li><p><a class="reference internal" href="#dna-protein-scaffolding-inpaint-2-dna-1-protein">Example 5: DNA–protein scaffolding, inpaint two DNA chains and one protein chain</a></p></li>
<li><p><a class="reference internal" href="#dna-origami-symmetric-denoising">Example 6: DNA origami with symmetric denoising</a></p></li>
<li><p><a class="reference internal" href="#rna-design-triple-helix">Example 7: RNA design with triple helix</a></p></li>
<li><p><a class="reference internal" href="#control-rna-tertiary-multi-contact">Example 8: Control of RNA tertiary structure with multi-contact specification</a></p></li>
<li><p><a class="reference internal" href="#pseudocyclic-symmetry-base-pair-patterning">Example 9: Pseudocyclic symmetry using procedurally generated base-pair patterning</a></p></li>
<li><p><a class="reference internal" href="#de-novo-holliday-junctions">Example 10: De novo Holliday junctions using strand exchange</a></p></li>
<li><p><a class="reference internal" href="#sequence-specification-sequence-design">Example 11: Sequence specification and sequence design</a></p>
<ul class="simple">
<li><p><a class="reference internal" href="#unconditional-design-rna-protein">Unconditional design of RNA and protein</a></p></li>
</ul>
  <!-- - [Unconditional design of DNA and protein](#unconditional-design-dna-protein)-->
</li>
<li><p><a class="reference internal" href="#further-reading">Further Reading</a></p></li>
</ul>
</section>
<section id="example-1-unconditional-rna-generation">
<span id="unconditional-rna-generation"></span><h2>Example 1: Unconditional RNA generation<a class="headerlink" href="#example-1-unconditional-rna-generation" title="Link to this heading">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">apptainer</span> <span class="n">run</span> <span class="o">--</span><span class="n">nv</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">SE3nv</span><span class="o">.</span><span class="n">sif</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">RFDpoly</span><span class="o">/</span><span class="n">rf_diffusion</span><span class="o">/</span><span class="n">run_inference</span><span class="o">.</span><span class="n">py</span> \
<span class="o">--</span><span class="n">config</span><span class="o">-</span><span class="n">name</span><span class="o">=</span><span class="n">multi_polymer</span> \
<span class="n">diffuser</span><span class="o">.</span><span class="n">T</span><span class="o">=</span><span class="mi">50</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">num_designs</span><span class="o">=</span><span class="mi">5</span> \
<span class="s1">&#39;contigmap.contigs=[&quot;90&quot;]&#39;</span> \
<span class="n">contigmap</span><span class="o">.</span><span class="n">polymer_chains</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;rna&#39;</span><span class="p">]</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">output_prefix</span><span class="o">=</span><span class="s1">&#39;./demo_outputs/RNA_uncond_standard_settings&#39;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">ckpt_path</span><span class="o">=/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">RFDpoly</span><span class="o">/</span><span class="n">weights</span><span class="o">/</span><span class="n">train_session2024</span><span class="o">-</span><span class="mi">07</span><span class="o">-</span><span class="mi">08_1720455712</span><span class="n">_BFF_3</span><span class="mf">.00</span><span class="o">.</span><span class="n">pt</span>
</pre></div>
</div>
</section>
<section id="example-2-unconditional-design-of-one-protein-chain-and-two-dna-chains">
<span id="unconditional-protein-2-dna"></span><h2>Example 2: Unconditional design of one protein chain and two DNA chains<a class="headerlink" href="#example-2-unconditional-design-of-one-protein-chain-and-two-dna-chains" title="Link to this heading">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">apptainer</span> <span class="n">run</span> <span class="o">--</span><span class="n">nv</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">SE3nv</span><span class="o">.</span><span class="n">sif</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">RFDpoly</span><span class="o">/</span><span class="n">rf_diffusion</span><span class="o">/</span><span class="n">run_inference</span><span class="o">.</span><span class="n">py</span> \
<span class="o">--</span><span class="n">config</span><span class="o">-</span><span class="n">name</span><span class="o">=</span><span class="n">multi_polymer</span> \
<span class="n">diffuser</span><span class="o">.</span><span class="n">T</span><span class="o">=</span><span class="mi">50</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">num_designs</span><span class="o">=</span><span class="mi">5</span> \
<span class="s2">&quot;contigmap.contigs=[20\ 20\ 75]&quot;</span> \
<span class="s1">&#39;contigmap.polymer_chains=[&quot;dna&quot;,&quot;dna&quot;,&quot;protein&quot;]&#39;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">output_prefix</span><span class="o">=</span><span class="s1">&#39;./demo_outputs/DNA_prot_uncond_standard_settings&#39;</span>
<span class="n">inference</span><span class="o">.</span><span class="n">ckpt_path</span><span class="o">=/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">RFDpoly</span><span class="o">/</span><span class="n">weights</span><span class="o">/</span><span class="n">train_session2024</span><span class="o">-</span><span class="mi">07</span><span class="o">-</span><span class="mi">08_1720455712</span><span class="n">_BFF_3</span><span class="mf">.00</span><span class="o">.</span><span class="n">pt</span>
</pre></div>
</div>
</section>
<section id="example-3-rna-riboswitch-design-with-conditioning-from-eterna-puzzle-structure-string">
<span id="rna-riboswitch-conditioning-eterna"></span><h2>Example 3: RNA riboswitch design with conditioning from Eterna puzzle structure string<a class="headerlink" href="#example-3-rna-riboswitch-design-with-conditioning-from-eterna-puzzle-structure-string" title="Link to this heading">¶</a></h2>
<p>Puzzle source: <a class="reference external" href="https://eternagame.org/labs/13195459">JR Openknot4 Week6 4RZD</a></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">((((</span><span class="o">...............</span><span class="p">))))(((((((((</span><span class="o">........</span><span class="p">)))))))))</span><span class="o">.......</span><span class="p">(((((((</span><span class="o">........</span><span class="p">)))))))(((</span><span class="o">...........</span><span class="p">)))</span><span class="o">.....</span>
</pre></div>
</div>
<p>Because Hydra cannot use parentheses in command-line arguments, replace “(” and “)” with “5”
and “3”.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">apptainer</span> <span class="n">run</span> <span class="o">--</span><span class="n">nv</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">SE3nv</span><span class="o">.</span><span class="n">sif</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">rf_diffusion</span><span class="o">/</span><span class="n">run_inference</span><span class="o">.</span><span class="n">py</span> \
<span class="o">--</span><span class="n">config</span><span class="o">-</span><span class="n">name</span><span class="o">=</span><span class="n">multi_polymer</span> \
<span class="n">diffuser</span><span class="o">.</span><span class="n">T</span><span class="o">=</span><span class="mi">50</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">num_designs</span><span class="o">=</span><span class="mi">5</span> \
<span class="s1">&#39;contigmap.contigs=[&quot;100&quot;]&#39;</span> \
<span class="n">contigmap</span><span class="o">.</span><span class="n">polymer_chains</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;rna&#39;</span><span class="p">]</span> \
<span class="n">scaffoldguided</span><span class="o">.</span><span class="n">target_ss_string</span><span class="o">=</span><span class="s1">&#39;5555...............3333555555555........333333333.......5555555........3333333555...........333.....&#39;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">output_prefix</span><span class="o">=</span><span class="s1">&#39;./demo_outputs/RNA_eterna_cond_standard_settings&#39;</span>
<span class="n">inference</span><span class="o">.</span><span class="n">ckpt_path</span><span class="o">=/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">RFDpoly</span><span class="o">/</span><span class="n">weights</span><span class="o">/</span><span class="n">train_session2024</span><span class="o">-</span><span class="mi">07</span><span class="o">-</span><span class="mi">08_1720455712</span><span class="n">_BFF_3</span><span class="mf">.00</span><span class="o">.</span><span class="n">pt</span>
</pre></div>
</div>
<p>Try this out on more <a class="reference external" href="https://eternagame.org/challenges/11843006">Eterna puzzles</a>! To get the ss-string for a given puzzle, start playing, then
right-click → “Copy Structure” or choose it from the menu in the upper left corner.
<strong>BUT WAIT!</strong> What if you want to insert multiple pseudoknots into your design or insert them
into specific positions of certain chains? You can provide a list of secondary structure strings as
follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">scaffoldguided</span><span class="o">.</span><span class="n">target_ss_string_list</span><span class="o">=</span><span class="p">[</span>\<span class="s1">&#39;B1-90:.5555555555555555555..ffffff.33333555....333555....33333333333333333555555..tttttt.333333.</span><span class="se">\&#39;</span><span class="s1">,</span><span class="se">\&#39;</span><span class="s1">A116-205:.5555555555555555555..ffffff.33333555....333555....33333333333333333555555..tttttt.333333.</span><span class="se">\&#39;</span><span class="s1">] </span>
</pre></div>
</div>
<p>where the ranges specify chains and indices of pseudoknot insertion locations in the output contigs.</p>
</section>
<section id="example-4-dnaprotein-scaffolding-inpaint-two-dna-chains">
<span id="dna-protein-scaffolding-inpaint-2-dna"></span><h2>Example 4: DNA–protein scaffolding, inpaint two DNA chains<a class="headerlink" href="#example-4-dnaprotein-scaffolding-inpaint-two-dna-chains" title="Link to this heading">¶</a></h2>
<img src="_static/example_4_new.png" alt="Image of an example output that highlights each generated chain and connects it to the portion of the contig that describes it." width=700px>
<p>The argument <code class="docutils literal notranslate"><span class="pre">inference.ij_visible</span></code> controls which motifs listed in the contigs have their relative
orientations locked during inference (credit to <a class="reference external" href="https://scholar.google.com/citations?user=k6PhNDQAAAAJ&amp;amp;hl=en">Dr. David Juergens</a> for this intuitive system). In the following
contact map, lowercase letters denote motifs from the input PDB in the order they occur in the contigs
(not to be confused with uppercase chain IDs):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">10</span><span class="p">,</span><span class="n">D8</span><span class="o">-</span><span class="mi">13</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="n">B8</span><span class="o">-</span><span class="mi">13</span><span class="p">,</span><span class="mi">10</span>\ <span class="mi">10</span><span class="p">,</span><span class="n">B18</span><span class="o">-</span><span class="mi">23</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="n">D18</span><span class="o">-</span><span class="mi">23</span><span class="p">,</span><span class="mi">10</span>\ <span class="n">A1</span><span class="o">-</span><span class="mi">56</span><span class="p">,</span><span class="mi">0</span>\ <span class="n">C1</span><span class="o">-</span><span class="mi">56</span><span class="p">,</span><span class="mi">0</span>
   <span class="n">a</span><span class="o">.</span>      <span class="n">b</span><span class="o">.</span>           <span class="n">c</span><span class="o">.</span>       <span class="n">d</span><span class="o">.</span>         <span class="n">e</span><span class="o">.</span>       <span class="n">f</span><span class="o">.</span>
</pre></div>
</div>
<p>Motifs A1-56, B8-13, and B18-23 (b,c,e) form the first DNA+binder group “bce”, while C1-5,
D8-13, and D18-23 (a,d,f) form “adf”. The relative orientations of “bce” and “adf” can now vary
during inference.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">SE3nv</span><span class="o">.</span><span class="n">sif</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">rf_diffusion</span><span class="o">/</span><span class="n">run_inference</span><span class="o">.</span><span class="n">py</span> \
<span class="o">--</span><span class="n">config</span><span class="o">-</span><span class="n">name</span><span class="o">=</span><span class="n">multi_polymer</span> \
<span class="n">diffuser</span><span class="o">.</span><span class="n">T</span><span class="o">=</span><span class="mi">50</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">num_designs</span><span class="o">=</span><span class="mi">5</span> \
<span class="s2">&quot;contigmap.contigs=[&#39;10,D8-13,6,B8-13,10 10,B18-23,6,D18-23,10 A1-56,0 C1-56,0&#39;]&quot;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">ij_visible</span><span class="o">=</span><span class="s1">&#39;bce-adf&#39;</span> \
<span class="s2">&quot;contigmap.polymer_chains=[dna,dna,protein,protein]&quot;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">input_pdb</span><span class="o">=</span><span class="s1">&#39;path/to/RFDpoly/test_data/combo_DBP009_DBP010_DBP011_with_DNA_v2.pdb&#39;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">output_prefix</span><span class="o">=</span><span class="s1">&#39;./demo_outputs/DNA_binders_scaffolding_test1_standard_settings&#39;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">ckpt_path</span><span class="o">=/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">RFDpoly</span><span class="o">/</span><span class="n">weights</span><span class="o">/</span><span class="n">train_session2024</span><span class="o">-</span><span class="mi">07</span><span class="o">-</span><span class="mi">08_1720455712</span><span class="n">_BFF_3</span><span class="mf">.00</span><span class="o">.</span><span class="n">pt</span>
</pre></div>
</div>
</section>
<section id="example-5-dnaprotein-scaffolding-inpaint-two-dna-chains-and-one-protein-chain">
<span id="dna-protein-scaffolding-inpaint-2-dna-1-protein"></span><h2>Example 5: DNA–protein scaffolding, inpaint two DNA chains and one protein chain<a class="headerlink" href="#example-5-dnaprotein-scaffolding-inpaint-two-dna-chains-and-one-protein-chain" title="Link to this heading">¶</a></h2>
<img src="_static/example_5_new.png" alt="Image of an example output that highlights each generated chain and connects it to the portion of the contig that describes it." width=700px>
<p>Here, contigmap.polymer chains lists only three chains because binding proteins are merged into
one.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">SE3nv</span><span class="o">.</span><span class="n">sif</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">rf_diffusion</span><span class="o">/</span><span class="n">run_inference</span><span class="o">.</span><span class="n">py</span> \
<span class="o">--</span><span class="n">config</span><span class="o">-</span><span class="n">name</span><span class="o">=</span><span class="n">multi_polymer</span> \
<span class="n">diffuser</span><span class="o">.</span><span class="n">T</span><span class="o">=</span><span class="mi">50</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">num_designs</span><span class="o">=</span><span class="mi">5</span> \
<span class="s2">&quot;contigmap.contigs=[&#39;5,D8-13,2,B8-13,5 5,B18-23,2,D18-23,5 A1-52,90,C4-56,0&#39;]&quot;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">ij_visible</span><span class="o">=</span><span class="s1">&#39;bce-adf&#39;</span> \
<span class="n">contigmap</span><span class="o">.</span><span class="n">polymer_chains</span><span class="o">=</span><span class="p">[</span><span class="n">dna</span><span class="p">,</span><span class="n">dna</span><span class="p">,</span><span class="n">protein</span><span class="p">]</span> \
<span class="n">scaffoldguided</span><span class="o">.</span><span class="n">target_ss_pairs</span><span class="o">=</span><span class="s2">&quot;[&#39;A1-24,B1-24&#39;]&quot;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">input_pdb</span><span class="o">=</span><span class="s1">&#39;path/to/RFDpoly/rf_diffusion/test_data/combo_DBP009_DBP010_DBP011_with_DNA_v2.pdb&#39;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">output_prefix</span><span class="o">=</span><span class="s1">&#39;./demo_outputs/DNA_binders_scaffolding_test2_standard_settings&#39;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">ckpt_path</span><span class="o">=/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">RFDpoly</span><span class="o">/</span><span class="n">weights</span><span class="o">/</span><span class="n">train_session2024</span><span class="o">-</span><span class="mi">07</span><span class="o">-</span><span class="mi">08_1720455712</span><span class="n">_BFF_3</span><span class="mf">.00</span><span class="o">.</span><span class="n">pt</span>
</pre></div>
</div>
<p>The new argument scaffoldguided.target_ss_pairs enforces base-pairing of polymer ranges
(here A1-24 B1-24). Each corresponds to the two full DNA contigs (A,B) in the outputs, each of
length 24.</p>
</section>
<section id="example-6-dna-origami-with-symmetric-denoising">
<span id="dna-origami-symmetric-denoising"></span><h2>Example 6: DNA origami with symmetric denoising<a class="headerlink" href="#example-6-dna-origami-with-symmetric-denoising" title="Link to this heading">¶</a></h2>
<p>The scaffoldguided.target ss pairs argument specifies paired ranges in the design; both
ranges in each pair must have equal length. The first is 5→3, the second 3→5.</p>
<img src="_static/rna_polyhedra.png" alt="Example of an output of this calculation." width=400>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">SE3nv</span><span class="o">.</span><span class="n">sif</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">rf_diffusion</span><span class="o">/</span><span class="n">run_inference</span><span class="o">.</span><span class="n">py</span> \
<span class="o">--</span><span class="n">config</span><span class="o">-</span><span class="n">name</span><span class="o">=</span><span class="n">multi_polymer</span> \
<span class="n">diffuser</span><span class="o">.</span><span class="n">T</span><span class="o">=</span><span class="mi">50</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">num_designs</span><span class="o">=</span><span class="mi">5</span> \
<span class="s2">&quot;contigmap.contigs=[&#39;60 60 60 60&#39;]&quot;</span> \
<span class="s2">&quot;contigmap.polymer_chains=[dna,dna,dna,dna]&quot;</span> \
<span class="n">scaffoldguided</span><span class="o">.</span><span class="n">target_ss_pairs</span><span class="o">=</span><span class="s2">&quot;[&#39;A1-20,B1-20&#39;,&#39;A21-40,C21-40&#39;,&#39;A41-60,D41-60&#39;,&#39;B21-40,D21-40&#39;,&#39;B41-60,C41-60&#39;,&#39;C1-20,D1-20&#39;]&quot;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">symmetry</span><span class="o">=</span><span class="s1">&#39;d2&#39;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">output_prefix</span><span class="o">=</span><span class="s1">&#39;./demo_outputs/DNA_origami_standard_settings&#39;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">ckpt_path</span><span class="o">=/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">RFDpoly</span><span class="o">/</span><span class="n">weights</span><span class="o">/</span><span class="n">train_session2024</span><span class="o">-</span><span class="mi">07</span><span class="o">-</span><span class="mi">08_1720455712</span><span class="n">_BFF_3</span><span class="mf">.00</span><span class="o">.</span><span class="n">pt</span>
</pre></div>
</div>
</section>
<section id="example-7-rna-design-with-triple-helix">
<span id="rna-design-triple-helix"></span><h2>Example 7: RNA design with triple helix<a class="headerlink" href="#example-7-rna-design-with-triple-helix" title="Link to this heading">¶</a></h2>
<p>Just as in the previous example, we can use base-paired ranges of sequence to control RNA
topology. By default, paired regions are antiparallel, but specific orientations can be assigned (e.g., triple helices with parallel/antiparallel combinations).</p>
<img src="_static/rna_triple_helix.png" alt="Example of an output of this calculation." width=500px>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">SE3nv</span><span class="o">.</span><span class="n">sif</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">rf_diffusion</span><span class="o">/</span><span class="n">run_inference</span><span class="o">.</span><span class="n">py</span> \
<span class="o">--</span><span class="n">config</span><span class="o">-</span><span class="n">name</span><span class="o">=</span><span class="n">multi_polymer</span> \
<span class="n">diffuser</span><span class="o">.</span><span class="n">T</span><span class="o">=</span><span class="mi">50</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">num_designs</span><span class="o">=</span><span class="mi">5</span> \
<span class="n">contigmap</span><span class="o">.</span><span class="n">contigs</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;75&#39;</span><span class="p">]</span> \
<span class="s2">&quot;contigmap.polymer_chains=[rna]&quot;</span> \
<span class="n">scaffoldguided</span><span class="o">.</span><span class="n">target_ss_pairs</span><span class="o">=</span><span class="s2">&quot;[&#39;A5-20,A55-70&#39;,&#39;A55-70,A30-45&#39;]&quot;</span> \
<span class="n">scaffoldguided</span><span class="o">.</span><span class="n">target_ss_pair_ori</span><span class="o">=</span><span class="s2">&quot;[P,A]&quot;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">output_prefix</span><span class="o">=</span><span class="s1">&#39;./demo_outputs/Triple_helix_test&#39;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">ckpt_path</span><span class="o">=/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">RFDpoly</span><span class="o">/</span><span class="n">weights</span><span class="o">/</span><span class="n">train_session2024</span><span class="o">-</span><span class="mi">07</span><span class="o">-</span><span class="mi">08_1720455712</span><span class="n">_BFF_3</span><span class="mf">.00</span><span class="o">.</span><span class="n">pt</span>
</pre></div>
</div>
<p>Each orientation in the list scaffoldguided.target_ss_pair_ori corresponds to the element at
the same index in scaffoldguided.target_ss_pairs.</p>
</section>
<section id="example-8-control-of-rna-tertiary-structure-with-multi-contact-specification">
<span id="control-rna-tertiary-multi-contact"></span><h2>Example 8: Control of RNA tertiary structure with multi-contact specification<a class="headerlink" href="#example-8-control-of-rna-tertiary-structure-with-multi-contact-specification" title="Link to this heading">¶</a></h2>
<img src="_static/rna_triple_tip.png" alt="Example of a the tertiary structure that can be produced with this calculation." width=400px>
<p>We can “staple” distal loops in RNA pseudoknots together by specifying regions of multi-base
contacts using scaffoldguided.force_multi_contacts. Secondary-structure strings cannot encode
beyond simple two-base pair configurations, so this feature enables higher-order tertiary interactions.
We can also force loop placement via scaffoldguided.force_loops_list.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">SE3nv</span><span class="o">.</span><span class="n">sif</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">rf_diffusion</span><span class="o">/</span><span class="n">run_inference</span><span class="o">.</span><span class="n">py</span> \
<span class="o">--</span><span class="n">config</span><span class="o">-</span><span class="n">name</span><span class="o">=</span><span class="n">multi_polymer</span> \
<span class="n">diffuser</span><span class="o">.</span><span class="n">T</span><span class="o">=</span><span class="mi">50</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">num_designs</span><span class="o">=</span><span class="mi">5</span> \
<span class="s2">&quot;contigmap.contigs=[&#39;80&#39;]&quot;</span> \
<span class="s2">&quot;contigmap.polymer_chains=[rna]&quot;</span> \
<span class="n">scaffoldguided</span><span class="o">.</span><span class="n">target_ss_pairs</span><span class="o">=</span><span class="s2">&quot;[&#39;A5-15,A25-35&#39;,&#39;A45-55,A65-75&#39;]&quot;</span> \
<span class="n">scaffoldguided</span><span class="o">.</span><span class="n">force_multi_contacts</span><span class="o">=</span><span class="s2">&quot;[&#39;A19,A61,A20&#39;,&#39;A59,A21,A60&#39;]&quot;</span> \
<span class="n">scaffoldguided</span><span class="o">.</span><span class="n">force_loops_list</span><span class="o">=</span><span class="s2">&quot;[&#39;A38-42&#39;]&quot;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">output_prefix</span><span class="o">=</span><span class="s1">&#39;./demo_outputs/loop_touch_test&#39;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">ckpt_path</span><span class="o">=/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">RFDpoly</span><span class="o">/</span><span class="n">weights</span><span class="o">/</span><span class="n">train_session2024</span><span class="o">-</span><span class="mi">07</span><span class="o">-</span><span class="mi">08_1720455712</span><span class="n">_BFF_3</span><span class="mf">.00</span><span class="o">.</span><span class="n">pt</span>
</pre></div>
</div>
</section>
<section id="example-9-pseudocyclic-symmetry-using-procedurally-generated-base-pair-patterning">
<span id="pseudocyclic-symmetry-base-pair-patterning"></span><h2>Example 9: Pseudocyclic symmetry using procedurally generated base-pair patterning<a class="headerlink" href="#example-9-pseudocyclic-symmetry-using-procedurally-generated-base-pair-patterning" title="Link to this heading">¶</a></h2>
<p>Below are arguments used to create pseudo-symmetry in a single chain forming a cyclic-symmetric shape (e.g., C2 pseudocycle).</p>
<img src="_static/c2_pseudocycle_png.png" alt="Example of a c2 pseudocycle." width=700px>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">SE3nv</span><span class="o">.</span><span class="n">sif</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">rf_diffusion</span><span class="o">/</span><span class="n">run_inference</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">config</span><span class="o">-</span><span class="n">name</span><span class="o">=</span><span class="n">multi_polymer</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">ckpt_path</span><span class="o">=</span><span class="s1">&#39;/path/to.RFDpoly/weights/train_session2024-07-08_1720455712_BFF_3.00.pt&#39;</span> \
<span class="n">diffuser</span><span class="o">.</span><span class="n">T</span><span class="o">=</span><span class="mi">50</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">num_designs</span><span class="o">=</span><span class="mi">5</span> \
<span class="s2">&quot;contigmap.contigs=[&#39;240&#39;]&quot;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">pseudo_symmetry</span><span class="o">=</span><span class="s1">&#39;c2&#39;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">n_repeats</span><span class="o">=</span><span class="mi">2</span> \
<span class="n">scaffoldguided</span><span class="o">.</span><span class="n">target_ss_pairs</span><span class="o">=</span><span class="s2">&quot;[&#39;A4-5,A237-238&#39;,&#39;A7,A236&#39;,&#39;A11-12,A230-231&#39;,&#39;A18-27,A137-146&#39;,&#39;A44-47,A72-75&#39;,&#39;A49-44,A64-69&#39;,&#39;A63-69,A49-55&#39;,&#39;A72-75,A44-47&#39;,&#39;A90-99,A197-206&#39;,&#39;A110-112,A130-132&#39;,&#39;A115-115,A127-128&#39;,&#39;A117-118,A124-125&#39;,&#39;A124-125,A117-118&#39;,&#39;A127-128,A115-116&#39;,&#39;A130-132,A110-112&#39;,&#39;A137-140,A24-27&#39;,&#39;A141-146,A18-23&#39;,&#39;A164-168,A191-195&#39;,&#39;A169-175,A183-189&#39;,&#39;A183-189,A169-175&#39;,&#39;A191-195,A164-168&#39;,&#39;A197-206,A90-99&#39;,&#39;A210-214,A82-86&#39;,&#39;A215-219,A77-81&#39;,&#39;A230-231,A11-12&#39;,&#39;A236,A7&#39;,&#39;A237-238,A4-5&#39;]&quot;</span> \
<span class="n">contigmap</span><span class="o">.</span><span class="n">polymer_chains</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;rna&#39;</span><span class="p">]</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">output_prefix</span><span class="o">=</span><span class="s1">&#39;./outputs_2025-02-03/pC2_test01__BFF_3.00&#39;</span> \
</pre></div>
</div>
</section>
<section id="example-10-de-novo-holliday-junctions-using-strand-exchange">
<span id="de-novo-holliday-junctions"></span><h2>Example 10: De novo Holliday junctions using strand exchange<a class="headerlink" href="#example-10-de-novo-holliday-junctions-using-strand-exchange" title="Link to this heading">¶</a></h2>
<img src="_static/holliday_junc_png.png" alt="Example of a designed Holliday junction." width=700px>
<p>We can use symmetry and strand exchange to design Holliday-junction-style complexes. Chain
and index specifications in <code class="docutils literal notranslate"><span class="pre">scaffoldguided.target_ss_pairs</span></code> refer to chain IDs and indices in the
output structure defined by contig topology.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">SE3nv</span><span class="o">.</span><span class="n">sif</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">rf_diffusion</span><span class="o">/</span><span class="n">run_inference</span><span class="o">.</span><span class="n">py</span> \
<span class="o">--</span><span class="n">config</span><span class="o">-</span><span class="n">name</span><span class="o">=</span><span class="n">multi_polymer</span> \
<span class="n">diffuser</span><span class="o">.</span><span class="n">T</span><span class="o">=</span><span class="mi">50</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">symmetry</span><span class="o">=</span><span class="s1">&#39;c2&#39;</span> \
<span class="n">contigmap</span><span class="o">.</span><span class="n">inpaint_seq</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;D60&#39;</span><span class="p">,</span><span class="s1">&#39;D57&#39;</span><span class="p">,</span><span class="s1">&#39;D58&#39;</span><span class="p">,</span><span class="s1">&#39;D11&#39;</span><span class="p">,</span><span class="s1">&#39;D15&#39;</span><span class="p">,</span><span class="s1">&#39;D19&#39;</span><span class="p">,</span><span class="s1">&#39;D8&#39;</span><span class="p">,</span><span class="s1">&#39;D61&#39;</span><span class="p">,</span><span class="s1">&#39;D18&#39;</span><span class="p">,</span><span class="s1">&#39;A9&#39;</span><span class="p">,</span><span class="s1">&#39;A59&#39;</span><span class="p">,</span><span class="s1">&#39;A62&#39;</span><span class="p">,</span><span class="s1">&#39;A8&#39;</span><span class="p">,</span><span class="s1">&#39;A4&#39;</span><span class="p">,</span><span class="s1">&#39;A5&#39;</span><span class="p">]</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">num_designs</span><span class="o">=</span><span class="mi">5</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">ckpt_path</span><span class="o">=</span><span class="s1">&#39;/path/to/RFDpoly/weights/train_session2024-07-08_1720455712_BFF_3.00.pt&#39;</span> \
<span class="s2">&quot;contigmap.contigs=[&#39;A1-61,60,D14-65 15,B6-12,4,F1-8,15 15,E7-14,4,C4-10,15 A1-61,60,D14-65 15,B6-12,4,F1-8,15 15,E7-14,4,C4-10,15&#39;]&quot;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">ij_visible</span><span class="o">=</span><span class="s1">&#39;acf-bde-gil-hjk&#39;</span> \
<span class="s2">&quot;contigmap.polymer_chains=[&#39;protein&#39;,&#39;dna&#39;,&#39;dna&#39;,&#39;protein&#39;,&#39;dna&#39;,&#39;dna&#39;]&quot;</span> \
<span class="n">scaffoldguided</span><span class="o">.</span><span class="n">target_ss_pairs</span><span class="o">=</span><span class="s2">&quot;[&#39;B1-10,F40-49&#39;,&#39;B40-49,F1-10&#39;,&#39;B16-34,C16-34&#39;,&#39;E16-34,F16-34&#39;,&#39;C1-10,E40-49&#39;,&#39;C40-49,E1-10&#39;]&quot;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">input_pdb</span><span class="o">=</span><span class="s1">&#39;/path/to/RFDpoly/rf_diffusion/test_data/DBP35opt_DBP48.pdb&#39;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">output_prefix</span><span class="o">=</span><span class="s1">&#39;./outputs_2025-02-03/DBP_scaffolding_test06__BFF_3.00&#39;</span>
</pre></div>
</div>
</section>
<section id="example-11-sequence-specification-and-sequence-design">
<span id="sequence-specification-sequence-design"></span><h2>Example 11: Sequence specification and sequence design<a class="headerlink" href="#example-11-sequence-specification-and-sequence-design" title="Link to this heading">¶</a></h2>
<p>We can now specify the sequence of our
structures to be whatever we want!
Additionally, the model has been trained to do sequence prediction, so we can decode a sequence during
the denoising trajectory (this allows us to generate outputs with all of the base atoms rendered for
NA design, as well as nice sidechain interactions for protein design).
The default behavior throughout RFdiffusion is to keep the sequence of diffused regions masked
during the trajectory, even if the outputs seem to have residue labels. RoseTTAfold must “see”
sequence labels in order to generate sidechains, so I added a flag, inference.update_seq_t=True,
which allows the model to see either a user-specified sequence or the model’s predicted sequence from
the previous timestep. This gives us sidechains, and is a unique addition to RFdiffusion-based tools.
There are two ways to control the sequence info during a trajectory:</p>
<ul class="simple">
<li><p>Turn on full-sequence visibility at some timestep towards the end of the trajectory, using inference.show_seq under t=15.</p></li>
<li><p>Gradually decode a random selection of positions at each step below some point in the trajectory, using diffuser.aa_decode_steps=40.
Both methods work nicely, so test them both during your design process. Examples using both methods are shown below.</p></li>
</ul>
<section id="example-11a-unconditional-design-of-rna-and-protein">
<span id="unconditional-design-rna-protein"></span><h3>Example 11A: Unconditional design of RNA and protein<a class="headerlink" href="#example-11a-unconditional-design-of-rna-and-protein" title="Link to this heading">¶</a></h3>
<p>While specifying the RNA sequence, letting the model design the protein sequence, and gradually revealing various sequence positions over the course of the last 40 steps:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">SE3nv</span><span class="o">.</span><span class="n">sif</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">rf_diffusion</span><span class="o">/</span><span class="n">run_inference</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">config</span><span class="o">-</span><span class="n">name</span><span class="o">=</span><span class="n">multi_polymer</span> \
<span class="n">diffuser</span><span class="o">.</span><span class="n">T</span><span class="o">=</span><span class="mi">50</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">num_designs</span><span class="o">=</span><span class="mi">3</span> \
<span class="s2">&quot;contigmap.contigs=[&#39;43 20 75&#39;]&quot;</span> \
<span class="s2">&quot;contigmap.polymer_chains=[&#39;rna&#39;,&#39;rna&#39;,&#39;protein&#39;]&quot;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">set_sequence</span><span class="o">=</span><span class="s2">&quot;[&#39;A1-43:GGAUGUACUACCAGCUGAUGAGUCCCAAAUAGGACGAAACGCC&#39;,&#39;B1-20:GGCGUCCUGGUAUCCAAUCC&#39;]&quot;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">update_seq_t</span><span class="o">=</span><span class="kc">True</span> \
<span class="n">diffuser</span><span class="o">.</span><span class="n">aa_decode_steps</span><span class="o">=</span><span class="mi">40</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">ckpt_path</span><span class="o">=</span><span class="s1">&#39;/path/to/RFDpoly/weights/train_session2024-07-08_1720455712_BFF_3.00.pt&#39;</span> \
<span class="n">inference</span><span class="o">.</span><span class="n">output_prefix</span><span class="o">=</span><span class="s1">&#39;./demo_outputs/RNA-prot_seq-spec_and_seq-design_standard_settings&#39;</span>
</pre></div>
</div>
<!-- (unconditional-design-dna-protein)=
### Example 11B: Unconditional design of DNA and protein
While specifying the DNA sequence, letting the model design the protein sequence, and letting the model see for the last 15 steps (currently throws an error but works with autoregressive decoding):
```
/path/to/SE3nv.sif /path/to/rf_diffusion/run_inference.py --config-name=multi_polymer \
diffuser.T=50 \
inference.num_designs=3 \
"contigmap.contigs=['33 33 75']" \
"contigmap.polymer_chains=['dna','dna','protein']" \
scaffoldguided.target_ss_pairs="['A11-23,B11-23']" \
inference.set_sequence="['A11-23:TAGCAGGATGTGT']" \
inference.assume_canonical_pair_seq=True \
inference.update_seq_t=True \
inference.show_seq_under_t=15 \
inference.output_prefix='./demo_outputs/DNA-prot_seq-spec_and_seq-design_standard_settings'
```
Notice in the DNA example above, only one of the dsDNA chains is specified. Since we specified
the paired regions using scaffoldguided.target ss pairs=['A11-23,B11-23'], the model knows
which bases should be paired. Then, we can use the flag inference.assume canonical pair seq=True
to fill in canonical base-pair partners automatically.-->
</section>
</section>
<section id="further-reading">
<span id="id1"></span><h2>Further Reading<a class="headerlink" href="#further-reading" title="Link to this heading">¶</a></h2>
<p>We can use several symbol pairs to denote which of two bases being paired is 5’-led or 3’-led. Unfortu-
nately, due to Hydra’s constraints on command-line tokens, we cannot use standard DBN conventions
and must exclude the following symbols: (, ), [, ], {, }, &lt;, &gt;. However, the following symbols can be
used instead to control hierarchical pseudoknot specification:</p>
<div class="table-wrapper colwidths-auto docutils container">
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>5’-led</p></th>
<th class="head"><p>3’-led</p></th>
<th class="head"><p>Explanation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>5</p></td>
<td><p>3</p></td>
<td><p>(5’→3’ and 3’→5’)</p></td>
</tr>
<tr class="row-odd"><td><p>i</p></td>
<td><p>j</p></td>
<td><p>(index convention)</p></td>
</tr>
<tr class="row-even"><td><p>f</p></td>
<td><p>t</p></td>
<td><p>(“from” and “to”)</p></td>
</tr>
<tr class="row-odd"><td><p>b</p></td>
<td><p>e</p></td>
<td><p>(“begin” and “end”)</p></td>
</tr>
<tr class="row-even"><td><p>A</p></td>
<td><p>a</p></td>
<td><p>(“begin” and “end”)</p></td>
</tr>
</tbody>
</table>
</div>
<p>The code accepts any of these symbols to denote base pairs using the dot-bracket-notation paradigm.
However, Hydra cannot parse parentheses, braces, or brackets in command-line arguments. If your
topology requires multiple distinct open/close pairs (as with pseudoknots), simply replace symbols as
long as they remain distinct.</p>
<section id="polymer-contigs">
<h3>Polymer Contigs<a class="headerlink" href="#polymer-contigs" title="Link to this heading">¶</a></h3>
<p>If you’ve gotten this far, you’ve probably already seen the arguments such as:
<code class="docutils literal notranslate"><span class="pre">contigmap.polymer_chains=&quot;['dna','dna','protein']&quot;</span></code>
To control polymer classes generated, specify the polymer type for each chain in the contig line;
otherwise, the model makes assumptions that may yield odd results (e.g., threading nucleic backbones
through protein motifs). Future updates may automate this behavior, but for now specify polymer
chains manually.</p>
</section>
<section id="a-note-on-the-different-contigs">
<h3>A Note on the Different Contigs<a class="headerlink" href="#a-note-on-the-different-contigs" title="Link to this heading">¶</a></h3>
<p>Different configs and their intended use cases:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">multi_polymer</span></code>: main config; handles motif scaffolding and secondary-structure conditioning.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">RNA_ss_cond</span></code>: similar to above, but loads a checkpoint fine-tuned on more RNA samples.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">RNA_uncond</span></code>: older configuration predating SS-conditioning. Produces diverse unconditional
designs but with less designable backbones.</p></li>
</ul>
<p>In practice, <code class="docutils literal notranslate"><span class="pre">multi_polymer</span></code> works best across all polymer classes and tasks — truly a “one-model-
to-rule-them-all” scenario.</p>
</section>
<section id="a-note-on-checkpoints">
<h3>A Note on Checkpoints<a class="headerlink" href="#a-note-on-checkpoints" title="Link to this heading">¶</a></h3>
<p>Many checkpoints are available, but the multi polymer config generally suffices. Alternative checkpoints can be found at:
<code class="docutils literal notranslate"><span class="pre">/rf_diffusion/config/inference</span></code></p>
</section>
<section id="arguments-that-specify-residues">
<h3>Arguments that Specify Residues<a class="headerlink" href="#arguments-that-specify-residues" title="Link to this heading">¶</a></h3>
<p>Chain–resi pairs (e.g., “A23”) specify that something should
happen to a given residue. Sometimes we use this to refer to the chain–resi pairing in the input PDB
(for example, in the <code class="docutils literal notranslate"><span class="pre">contigmap.contigs</span></code> argument). Other times we refer to a location within the
newly generated structure (for example, in <code class="docutils literal notranslate"><span class="pre">scaffoldguided.target_ss_pairs</span></code>).</p>
</section>
<section id="a-note-on-pymol-s-cartoon-representation">
<h3>A Note on PyMOL’s Cartoon Representation<a class="headerlink" href="#a-note-on-pymol-s-cartoon-representation" title="Link to this heading">¶</a></h3>
<p>Since RFDpoly is based on the original implementation of RFdiffusion, its outputs lack sidechains in non-motif (designed) regions. This is particularly problematic for nucleic acids: PyMOL refuses to render nucleic acid cartoons if sidechains are missing. Therefore, Andrew recommends inspecting outputs in ChimeraX or using an alternative PyMOL scheme to visualize NA outputs.</p>
<p>If outputs seem invisible in PyMOL, it’s not a bug, just a rendering limitation.</p>
</section>
<section id="questions">
<h3>Questions?<a class="headerlink" href="#questions" title="Link to this heading">¶</a></h3>
<p>If you have any questions about RFDpoly, open a Discussion on GitHub or post in the Rosetta Commons Slack (Commons members only). If you have issues running any of the examples, open an Issue on GitHub.</p>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          <a class="prev-page" href="installation_guide.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Installation Guide</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2025, Favor, Andrew H and Quijano, Riley and Chernova, Elizaveta and Kubaney, Andrew and Weidle, Connor and Esler, Morgan A and McHugh, Lilian and Carr, Ann and Hsia, Yang and Juergens, David and others
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">RFDpoly usage examples</a><ul>
<li><a class="reference internal" href="#table-of-contents">Table of Contents</a></li>
<li><a class="reference internal" href="#example-1-unconditional-rna-generation">Example 1: Unconditional RNA generation</a></li>
<li><a class="reference internal" href="#example-2-unconditional-design-of-one-protein-chain-and-two-dna-chains">Example 2: Unconditional design of one protein chain and two DNA chains</a></li>
<li><a class="reference internal" href="#example-3-rna-riboswitch-design-with-conditioning-from-eterna-puzzle-structure-string">Example 3: RNA riboswitch design with conditioning from Eterna puzzle structure string</a></li>
<li><a class="reference internal" href="#example-4-dnaprotein-scaffolding-inpaint-two-dna-chains">Example 4: DNA–protein scaffolding, inpaint two DNA chains</a></li>
<li><a class="reference internal" href="#example-5-dnaprotein-scaffolding-inpaint-two-dna-chains-and-one-protein-chain">Example 5: DNA–protein scaffolding, inpaint two DNA chains and one protein chain</a></li>
<li><a class="reference internal" href="#example-6-dna-origami-with-symmetric-denoising">Example 6: DNA origami with symmetric denoising</a></li>
<li><a class="reference internal" href="#example-7-rna-design-with-triple-helix">Example 7: RNA design with triple helix</a></li>
<li><a class="reference internal" href="#example-8-control-of-rna-tertiary-structure-with-multi-contact-specification">Example 8: Control of RNA tertiary structure with multi-contact specification</a></li>
<li><a class="reference internal" href="#example-9-pseudocyclic-symmetry-using-procedurally-generated-base-pair-patterning">Example 9: Pseudocyclic symmetry using procedurally generated base-pair patterning</a></li>
<li><a class="reference internal" href="#example-10-de-novo-holliday-junctions-using-strand-exchange">Example 10: De novo Holliday junctions using strand exchange</a></li>
<li><a class="reference internal" href="#example-11-sequence-specification-and-sequence-design">Example 11: Sequence specification and sequence design</a><ul>
<li><a class="reference internal" href="#example-11a-unconditional-design-of-rna-and-protein">Example 11A: Unconditional design of RNA and protein</a></li>
</ul>
</li>
<li><a class="reference internal" href="#further-reading">Further Reading</a><ul>
<li><a class="reference internal" href="#polymer-contigs">Polymer Contigs</a></li>
<li><a class="reference internal" href="#a-note-on-the-different-contigs">A Note on the Different Contigs</a></li>
<li><a class="reference internal" href="#a-note-on-checkpoints">A Note on Checkpoints</a></li>
<li><a class="reference internal" href="#arguments-that-specify-residues">Arguments that Specify Residues</a></li>
<li><a class="reference internal" href="#a-note-on-pymol-s-cartoon-representation">A Note on PyMOL’s Cartoon Representation</a></li>
<li><a class="reference internal" href="#questions">Questions?</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="_static/documentation_options.js?v=8d563738"></script>
    <script src="_static/doctools.js?v=fd6eb6e6"></script>
    <script src="_static/sphinx_highlight.js?v=6ffebe34"></script>
    <script src="_static/scripts/furo.js?v=46bd48cc"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    </body>
</html>